meps/mep-0004.md

```diff
@@ -0,0 +1,273 @@
+
+## Proposal
+
+### AST Structure Changes
+
+The MyST AST will be extended to support per-output AST representation through the following changes:
+
+**Previous Structure (Version 2):**
+
+```typescript
+type OutputV2 = {
+  type: 'output';
+  data?: any[]; // Array of IOutput bundles
+  visibility?: any;
+};
+```
+
+In this structure, a single `Output` node contained an **array** of output data bundles in the `data` field.
+
+**New Structure (Version 3):**
+
+```typescript
+type Output = {
+  type: 'output';
+  children: (FlowContent | ListContent | PhrasingContent)[];
+  jupyter_data: IOutput; // Single IOutput bundle from Jupyter Notebooks
+};
+
+type Outputs = {
+  type: 'outputs';
+  children: (Output | FlowContent | ListContent | PhrasingContent)[]; // Support placeholders in addition to outputs
+  visibility?: Visibility; // `show`, `hide`, or `remove`
+  scroll?: boolean;
+};
+```
+
+The key changes are:
+
+1. **`Output` nodes** now represent a single output with its own AST subtree in `children` and a single `jupyter_data` field (instead of an array).
+
+2. **`Outputs` node** is introduced as a container with `Output` children, where there is a 1:1 correspondence between `Output` children and `IOutput` bundles.
+
+3. Each `Output` child can have its own AST subtree, enabling per-output parsing and reference resolution.
+
+### Migration Strategy
+
+Existing identifiers for the `Output` / `Outputs` nodes are not modified, as these are considered "content". The migration path includes:
+
+**Upgrade (V2 → V3):**
+
+- Convert each `Output` node with `data` array into an `Outputs` node
+- Create one `Output` child per item in the original `data` array
+- Distribute the original `children` to the first output if there is only one output
+- Preserve placeholders and other special children
+
+**Downgrade (V3 → V2):**
+
+- Convert `Outputs` nodes back to `Output` nodes
+- Collect all `jupyter_data` fields from `Output` children into a `data` array
+- Merge all `Output` children AST subtrees into a single `children` array
+- Preserve identifiers and other metadata
+
+### Example Transformation
+
+**Before (V2):**
+
+```json
+{
+  "type": "output",
+  "data": [
+    { "output_type": "stream", "text": "Hello" },
+    { "output_type": "display_data", "data": { "text/markdown": "**Bold**" } }
+  ],
+  "children": [{ "type": "text", "value": "Shared content" }]
+}
+```
+
+**After (V3):**
+
+```json
+{
+  "type": "outputs",
+  "children": [
+    {
+      "type": "output",
+      "jupyter_data": { "output_type": "stream", "text": "Hello" },
+      "children": []
+    },
+    {
+      "type": "output",
+      "jupyter_data": { "output_type": "display_data", "data": { "text/markdown": "**Bold**" } },
+      "children": [
+        {
+          "type": "strong",
+          "children": [{ "type": "text", "value": "Bold" }]
+        }
+      ]
+    }
+  ]
+}
+```
+
+## Examples
```

Comment:
I think we should produce multiple MEPs. In this one, it's just AST reworking. 

We should remove this - it's not part of *this* MEP. Well make another MEP to handle outputs parsing.